### ISSUE TITLE
Why we need two datasets and two autoencoders?

### ISSUE THREAD

**Issue by xthan** [deepfakes/faceswap#229]
Hi,

I do not understand why we need two folders containing faces of two people to train the faceswap model. Since the loss is simply a reconstruction error between a warped face and an unwarped face of the same person: 
https://github.com/deepfakes/faceswap/blob/6bea4c2b276f105624e18793082734f31c683743/lib/ModelAE.py#L61-L62

I think we can train one encoder + one decoder (one autoencoder) for a person (e.g., Trump), and during testing, just apply this encoder-decoder model to any face (no need to be Cage's) and it will convert it to Trump's face, right?

Similar idea is present in 
https://arxiv.org/pdf/1611.09577.pdf

**Comment by GokulEpiphany** [comment]
Your model would work well if the input had only one face. Won't the single autoencoder swap out all the faces in the input if it had more than one face?

**Comment by iperov** [comment]
actually there is ONE Encoder training on two decoders.

```
self.encoder = self.Encoder()
self.decoder_A = self.Decoder()
self.decoder_B = self.Decoder()
...
self.autoencoder_A = KerasModel(x, self.decoder_A(self.encoder(x)))
self.autoencoder_B = KerasModel(x, self.decoder_B(self.encoder(x)))
```

**Comment by Clorr** [comment]
@xthan I don't know if there is a strong theory behind the faceswap here. Here are some links about it:
- https://github.com/shaoanlu/faceswap-GAN/issues/26
- https://github.com/deepfakes/faceswap/issues/73

As @iperov said, we train a common encoder and 2 decoders (and we don't use the FaceA decoder when swapping FaceA to FaceB).

The purpose of that is to have an encoder that outputs good features when he sees a FaceA image, and that FaceB decoder unsderstands these features.

Schematically, I would summarize it like this:
- Encoder learns to recognize both FaceA and FaceB, and decoders, trained at the same time, know how to interpret encoder's output
- When we convert faces, encoder recognizes a FaceA, and generates an output that decoder of FaceB will understand

If we train the model on just one face, this would go like this:
- Encoder recognizes only FaceB and maybe overfit on FaceB
- When we try to faceswap FaceA, encoder will not recognize FaceA, and will generate an output that decoder does not understand. So it gives bad result.

On my side, I'm trying to make a generic encoder that is train with many faces so that he is not trained just on FaceB, but on thousands of different faces, so that it should output generic face features and won't be too overfitted to FaceB. Once done, I will try to train just a decoder with FaceB. @shaoanlu tried something like this already in [its One-For-All attempt](https://github.com/shaoanlu/faceswap-GAN/tree/master/notes). He was not totally satisfied of the result, but I think it is already interesting...

**Comment by Clorr** [comment]
I close this as it should be opened in https://github.com/deepfakes/faceswap-model

**Comment by KeyKy** [comment]
@Clorr So, have you successfully trained a generic encoder??
